Expose build.target .cargo/config setting as packages.target in Cargo.toml

Ekleog opened the pull request:
Hey!

I'm trying to do my first cargo contribution by implementing per-crate target settings as per [the irlo thread](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336) ; and I think I have a draft that looks good-ish (the root units returned by `generate_targets` have the right kinds set).

Closes #7004

**_Edit: the below problem description is now solved in the latest version of this PR, please ignore_**

But for some reason running on a test project now blocks on `Blocking waiting for file lock on build directory` and I have literally no idea how my changes could trigger this… would anyone have an idea of how the changes could lead to infinitely blocking there? (I already tried cargo clean just in case and it didn't appear to help)

FWIW, the output that looks hopeful to me is, on my testbed workspace:
```
Root units [out of generate_targets] are [...]:
 - package ‘smtp-client’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-server’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘yuubind-config’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-message-fuzz’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘yuubind-rpc’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘yuubind-config-example’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-message-fuzz’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘yuubind-config-example’ with kind ‘Target(CompileTarget { name: "wasm32-unknown-unknown" })’
 - package ‘smtp-queue’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-message-fuzz’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-message’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-server-fuzz’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘yuubind-config’ with kind ‘Target(CompileTarget { name: "wasm32-unknown-unknown" })’
 - package ‘yuubind’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
 - package ‘smtp-queue-fs’ with kind ‘Target(CompileTarget { name: "x86_64-unknown-linux-gnu" })’
```
(where both `yuubind-config` and `yuubind-config-example` are being configured to be `wasm32-unknown-unknown` and the other ones stay as host).

Interestingly enough, if I remove the `target` setting from `yuubind-config` (and leave it on `yuubind-config-example`) then it does no longer block on waiting for file lock on build directory, even though it does not actually compile with `wasm32-unknown-unknown`. And it does appear to correctly build yuubind-config-example as wasm32.

My investigation shows that it appears to happen iff there is a package with `package.target` being set that has dependencies.

This most likely is a bug in my code (eg. I build only the root units and not the whole unit graph maybe?), and am going to keep investigating it as such, but maybe someone would already know how dependency resolution could interact with build lock acquisition and give me hints?

Anyway, thank you all for all you do cargo!

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @ehuss (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

Ekleog commented:
As this PR currently appears to work (for my test repository), I'd like to ask for a first round of review before polishing up all the details that are sure to be problematic.

What do you think about this? As this is my first cargo contribution, maybe there is some obvious place I forgot changing, and I should be doing this before even thinking of adding tests and polishing things up? Also, should I (and if yes is there an example somewhere in the code of how) hide this new option behind an unstable gate?

If things look good enough, I'll be glad to poke around the cargo test framework and try adding tests, and to look for what documentation might need updating.

(BTW, there are already failing CI tests that appear to be unrelated to my changes, like [this one](https://github.com/rust-lang/cargo/pull/9030/checks?check_run_id=1628790153), I'm not sure whether I'm expected to drive-by fix them or not)

Eh2406 commented:
With the holidays in the US we are not paying as much attention as we normally do. For example the unrelated CI failing are something to do with Nightly, and normally there would be a preity prompt PR fixing it (or turning off Nightly testing) that you could rebase onto. Thank you for your contribution and patients.

Tests and documentation would be appreciated. I for one find it a good place to start finding out "what are you trying to do" as a start of a code review.
It is possible that we will make this insta stable, but it is more likely that when someone does a full review we will want a feature flag. Instructions for adding one are at https://github.com/rust-lang/cargo/blob/8917837fb682c6515c1cbfdbd16be95bb64bd6f1/src/cargo/core/features.rs

Eh2406 commented:
Thanks to @ehuss for #9033 that fixes CI.

Ekleog commented:
Thank you for your feedback! I have just fixed an issue I noticed while using a patched cargo, and rebased to a master that hopefully has CI fixed.

I also added the documentation, that hopefully should help understanding what this diff is about.

I haven't added tests yet, though, as I'd like to confirm that at least the idea of the change is legit. The irlo thread listed in the top post, AFAIU, says that this setting should be made available, but I'm not sure that eg. `[package]` is the right place to put it in.

Does this sound reasonable enough? And what do you think about [the documented changes](https://github.com/rust-lang/cargo/pull/9030/files#diff-3bc7476e0b88b61ea4e3bd46743021ef33c927f3052c1f7800bc8aa053627b31), that I hopefully implemented not too awfully here?

Cheers!

Ekleog commented:
Another thought I just came up with: should `package.target = "wasm32-unknown-unknown"` imply `lib.test = false`? What about `package.target = "i686-unknown-linux-gnu"` when the host is `x86_64-unknown-linux-gnu`?

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -672,13 +672,18 @@ impl RustcTargetData {
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| p.manifest().kind().into_iter()));
```
I think that this will need to iterate over the entire crate graph instead of just `ws.members()` to find requested targets?

Ekleog replied:
I've tried thinking about this, but it looks to me like `RustcTargetData` does not already have information enough to actually list the entire crate graph, and this only happens in `BuildContext` which gets called after that? And as `RustcTargetData` also gets used from eg. `cargo fetch`, which AFAIU does itself not have enough information to iterate over the entire crate graph — and it is indeed using the kind data in `RustcTargetData`.

Now, seeing as the structures filled here are AFAIU just caches, it'd be very possible to just fill them in on-demand instead, which should solve the issue. Would that sound good to you?

alexcrichton commented on src/cargo/core/compiler/compilation.rs:
```diff
@@ -116,11 +119,13 @@ impl<'cfg> Compilation<'cfg> {
-                .chain(Some(&CompileKind::Host))
+                .copied()
+                .chain(Some(CompileKind::Host))
+                .chain(ws.members().flat_map(|p| p.manifest().kind().into_iter()))
```
Similar to above, I think this needs to iterate over the entire crate graph?

(this may indicate that this should be refactored to a separate step where only one location determines all the targets in play and then that's passed as an argument here or stored in `bcx`)

alexcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -287,10 +287,20 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+            .requested_kinds
+            .iter()
+            .copied()
+            .chain(self.bcx.unit_graph.keys().map(|u| u.kind));
```
Ah yeah after reading this again, I think that `bcx` will definitely want to have a list (deduped) of all kinds in play for any given compilation.

alexcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -319,12 +329,17 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
-        for &kind in self
+        // TODO: is using `lto` as a list of all `Unit`s right here?
+        // That feels wrong, but it also looks like it's the least bad
+        // way of accessing that list from here
```
By storing all relevant targets in `bcx` I think we'd solve this.

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -857,7 +864,21 @@ fn generate_targets(
+            let explicit_kinds: Vec<_> = requested_kinds
+                .iter()
+                .map(|kind| match kind {
+                    CompileKind::Host => pkg.manifest().kind().unwrap_or(explicit_host_kind),
```
I'm not 100% sure on this interaction myself. For example if you have an application that embeds a wasm blob, then you probably want the wasm blob always compiled as wasm but the application itself may want to get compiled for different architectures. I could also see, though, use cases where `target` is the default but it should still be overridable with `--target`.

Basically I don't think there's a great answer at this time and we'll likely want to tweak the design here.

Ekleog replied:
Yes, this is also one thing I was wondering about. An idea I had was to add `default-target` and `forced-target` instead of adding `target`, and then it'd look like this:
```
pkg.manifest().forced_kind().unwrap_or(match kind {
    CompileKind::Host => pkg.manifest().default_kind().unwrap_or(explicit_host_kind),
    // ...
})
```

How would that sound? To me it feels like it's a bit more complex to use, so I was wondering which use case should be prioritized, and so for the initial draft I went with “just reuse the behavior from `.cargo/config` of making it a default” ; but it wasn't because I have found a good reason either way — actually now that I think about it more I kind of believe that if only one of default/forced should be made available, it might be better for it to be forced, as I can't think of good use cases where `target` should be a default while also being per-crate instead of in `.cargo/config`

alexcrichton replied:
Indeed yeah having both keys in the manifest would solve this issue. I'm not sure if it's the best design but this is all unstable anyway so we can tweak it later.

alexcrichton commented on src/cargo/util/toml/mod.rs:
```diff
@@ -807,6 +808,7 @@ pub struct TomlProject {
     authors: Option<Vec<String>>,
     build: Option<StringOrBool>,
     metabuild: Option<StringOrVec>,
+    target: Option<String>,
```
We'll want to gate this with a nightly manifest feature if it's set to `Some` to avoid insta-stabilizing this as well.

alexcrichton reviewed (commented):
Thanks for the PR! I think it may make sense to land a few of the refactorings in the backend (e.g. storing a list of kinds in the `BuildContext`) ahead of time before this PR if you'd like.

Otherwise I think it'd be good too add some tests here. Additionally I'm not entirely sure what the final design will be for `target` and how it interacts with `--target`, but I added some comments below on that.

Ekleog commented:
I think I have noticed another issue of this PR as is, though I still have to actually check it: feature resolution happens only once, and so if a crate is depended on with a different set of features by `target = "wasm32-unknown-unknown"` and host-target crates, it'll end up with all features enabled for both targets. Which can end up causing issues, when eg. one optional dependency isn't supported on one of the targets.

Fixing this would require changing the feature resolution algorithm, which I expect would be quite a bit more invasive. How do you think we should go? First land the “basic” stuff and then work towards fixing feature resolution while keeping the flag gated? Or finish work on getting the feature resolution first, and only then land it all at once, at the expense of this ending up being a (much?) bigger PR?

Ekleog commented:
Uh well, I posted that message faster than I wanted to, sorry for the notification noise.

So I also wanted to say, I've addressed a few of the comments, and replied to a few other comments. Split the refactoring towards #9046, and merged that branch here so it still works (hope git won't merge conflict me at the end, it's the first time I work in a “please don't rebase” scenario, usually I'd have just squash-rebased on top of that other branch)

Note to self about things still remaining:
- Fixing the `RustcTargetData` handling, maybe by filling the cache on-demand? (if that's confirmed to be the way to go)
- Feature-gate the two flags
- Add tests
- Figure out how to move forward with respect to feature flags that get depended on by multiple targets

Ekleog commented:
(Note: I've just rebased on top of master now that #9046 landed)

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -672,13 +672,23 @@ impl RustcTargetData {
         let mut target_info = HashMap::new();
-        for kind in requested_kinds {
-            if let CompileKind::Target(target) = *kind {
+        // Note that we cannot just rely on getting kinds for all units as the units aren't built yet (TODO CHECK THIS IS TRUE?)
```
Hm this may pose a bit of a chicken-and-egg problem where we may not be able to precalculate all target information ahead of time like we do today. This'll definitely  need to iterate over the entire dependency graph to discover targets since any package could introduce a new target I think.

Ekleog replied:
Do you think lazily filling the two `HashMaps` in the functions that actually access them could make sense?

Ekleog replied:
Hmm… Actually, I thought I'd just go ahead and implement this, but it turns out it's a much bigger refactoring than I initially thought. I've sent a WIP PR at https://github.com/rust-lang/cargo/pull/9049 ; if you think that's a good way to go I can finish it (already went up to the borrowck phase), but I'm starting to feel like this is not good design.

Another idea I had may be: add a function to `RustcTargetData` to manually request loading of more target information, and then call it from the places that may introduce new targets. What do you think about this other idea?

alexcrichton replied:
I don't have much of a strong preference for how this is fixed myself, I mostly just wanted to point out that this was a problem that needed fixing with this implementation.

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -857,7 +864,27 @@ fn generate_targets(
+                        CompileKind::Target(t) => CompileKind::Target(*t),
+                    })
+                    .collect()
+            };
```
I think that this may not actually be the right place for this conditional (or it may need to be in more than one location). For example if a build script depends on a wasm blob (for whatever reason) I don't think that's factored in here? 

I'm not sure if it's possible to easily expose this today really, but when we can depend on binary artifacts themselves (binaries, cdylibs, etc), we'll need to factor in manifest-requested targets to the `Unit` graph construction steps.

Ekleog replied:
I'm not sure I understand correctly here. Do you mean that, when binary artifact dependencies happen in `cargo`, then this code will need to be overhauled again to properly handle this case? 

Or is it already possible for a build script to depend on an artifact generated by another target, which would mean that there is an action to be undertaken now?

alexcrichton replied:
No I think you're right, without artifact dependencies this doesn't really matter. I think this is probably good enough for now on a second pass. This'll just need to be updated when we get artifact dependencies.

Ekleog commented:
Ok, so with this latest commit I have I think solved all the issues that have been raised, except:
1. Actually adding kinds after the fact, because I could not think of any case where the full list of kinds is not known by just exploring the workspace yet. This is because AFAIK artifact dependencies are not a thing yet — as soon as they are we'll just have to call `.add_kind()` at the places where testing shows crashes and things should just work. But I have not been able yet to think of any test case that would hit a situation where a kind only appears outside the workspace, as then the unit kind would AFAIU just be ignored to take its reverse-dep's kind instead.
2. Handling feature resolution by having one feature resolution per kind, to avoid that a crate that has features F and G, where F is used by wasm32-wasi and G by $host, ends up building both F and G for both wasm32-wasi and $host. (I have a concrete case where this happened, and F cannot build on $host nor G on wasm32-wasi) The solution to this needs to be informed by the current state of implementing artifact dependencies, as artifact dependencies will also influence feature resolution: if A depends on artifacts from B and C, B depends on D with feature F and C on D with feature G, do both B and C build with a D that has features F and G, or do they each get their own D? Depending on the response of this question + the status of the implementation of artifact dependencies, it will be hopefully possible to make progress here without creating merge conflicts between the two PRs — I don't want to cause trouble to other PRs that would currently be in the process of being written :)

bors commented:
:umbrella: The latest upstream changes (presumably #8037) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -694,13 +689,44 @@ impl RustcTargetData {
+        Ok(res)
+    }
+
+    pub fn add_kind(&mut self, kind: &CompileKind) -> CargoResult<()> {
```
Does this need to be a public method? It seems like it could probably be inlined above perhaps?

Ekleog replied:
The reason why it's `pub` is because it will make it easy to add kinds after-the-fact: just call `add_kind` at the places the kinds are added and that should be it.

Inlining in the above would certainly be possible — as far as I remember it's basically the way it was done before, and I've extracted this function in the process of adding kinds everywhere required before noticing that as we don't have artifact dependencies yet and there was thus no place to call this function from yet.

Do you want me to remove this function, or should I let it available so that it's easier to implement artifact dependencies when the time comes?

alexcrichton replied:
For now let's go ahead and leave it folded into the constructor above, we can always easily add it back later if necessary.

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -694,13 +689,44 @@ impl RustcTargetData {
+            self.target_info.insert(
+                *target,
+                TargetInfo::new(self.config, &self.requested_kinds, &self.rustc, *kind)?,
+            );
```
Since this might presumably be called for the same `target` multiple times could this avoid doing work if we alread know about `target`?

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -694,13 +689,44 @@ impl RustcTargetData {
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| {
```
My comments later about this needing to traverse the whole dependency graph I don't think are quite right until we have artifact dependencies. For now I think it should be safe to only traverse the workspace since that's the only location of the root units which is where targets can come from. Can you add a comment to this effect here though?

alexcrichton reviewed (commented):
I think it's ok to not support adding kinds after-the-fact, ideally we won't ever need to do that and I think that at least without artifact dependencies there's no need to do that just yet.

For feature resolution I think it's probably best to leave the nitty-gritty there as an open question on a tracking issue, we probably don't need full semantics of that here just yet. Does what's here not work in some situations, though?

As a final though, can you add some validation that `package.target` can only be used if the package has binary targets? For example it needs to have a dylib/cdylib/bin/etc. If it's just an rlib that should probably be an error.

alexcrichton commented:
Hm no nevermind scratch that about being an error, I could see it being useful still for just-libs that have tests and only want to `cargo check` on one target, for example.

Ekleog commented:
Thank you for your feedback! As there was a merge conflict, I've just added two commits to handle your review comments, and then rebased the whole thing on master, so the only two changes (apart from the few merge conflicts, particularly the feature-gate-format change in the new 1c515fe) are in the last two commits :)

Ekleog commented:
Oh and I just noticed I forgot to answer this question:
> For feature resolution I think it's probably best to leave the nitty-gritty there as an open question on a tracking issue, we probably don't need full semantics of that here just yet. Does what's here not work in some situations, though?

So there _is_ one situation I found where the current semantics are problematic: when one feature is incompatible with one target.

For instance, I have a feature that makes sense only on wasm32, in which I've put `const_assert_eq!(mem::size_of::<usize>(), mem::size_of::<u32>());` because I wouldn't be surprised if there were memory safety issues that could arise from some conversions in there.

With the current way feature resolution works (ie. one big fixpoint on the whole crate graph, regardless of build kind), if non-wasm32 depends on this crate and wasm32 code depends on this crate with the wasm32-specific feature enabled, then the non-wasm32 code will also try to build the wasm32-specific feature and will thus fail to build.

So, especially given it would probably be a breaking change to do that later (because we'd remove features that were previously enabled, though they were not explicitly depended upon… not sure what the actual policy is there, I'd guess it's OK if crates on crates.io remove a feature from one of their dependencies but it probably wouldn't be OK for cargo to do so), I think we will have to fix this before stabilization and I don't think there'd be much debate as to the expected semantics.

IOW, I'm mostly wondering whether it's better to do it just now, or to do it in a follow-up PR, to make things more manageable to review :) (though I must say I haven't implemented it yet so I don't know yet how hard it'll be to do)

alexcrichton commented:
Nah that's a really good point about target-specific features. Still ok to not handle here though, fine if we handle that later! I do think it will block stabilization of any form, however. All that being said I don't think it'll be the end of the world to soup-up the v2 feature resolver to fix this.

alexcrichton commented on crates/cargo-test-support/src/cross_compile.rs:
```diff
@@ -180,6 +181,51 @@ rustup does not appear to be installed. Make sure that the appropriate
+            return t;
+        }
+        let out = String::from_utf8(
+            cargo::util::process("rustc")
```
FWIW instead of executing `rustc -vV` could this use a build script to just forward the build scripts `env::var("TARGET")` to here?

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -688,13 +683,62 @@ impl RustcTargetData {
+            match self.target_config.entry(*target) {
+                std::collections::hash_map::Entry::Occupied(_) => (),
+                std::collections::hash_map::Entry::Vacant(place) => {
+                    place.insert(self.config.target_cfg_triple(target.short_name())?);
```
I think this can probably use `entry(...).or_insert_with(|| ..)` to avoid the `match`?

Ekleog replied:
Unfortunately no, `.or_insert_with(|| ...)` would not make it possible to pass the `?` out, as it doesn't make the closure fallible and we need it to be able to call `target.short_name`

danhoffman699 commented:
What's the status of this PR? No rush, but I'm eager to use this feature in nightly if possible.

Thanks!

bors commented:
:umbrella: The latest upstream changes (presumably #9184) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Ekleog commented:
I've been busy with work and life stuff over the past few weeks, sorry for not having been able to come back to this PR earlier! Here it is updated with the review comments handled — this idea of taking `TARGET` from build.rs is awesome!

I've pushed two new commits, and then force-pushed a rebase to fix the merge conflict. Hopefully this now looks good :)

I've prepared the text for the tracking issue by copying the formatting of a random other tracking issue, not sure whether it's supposed to be posted before or after the PR lands:

```
#### Summary

Feature gate: `per_package_target`
Discussion: https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336
Implementation: https://github.com/rust-lang/cargo/pull/9030
Documentation: TODO (based on [these previous changes](https://github.com/rust-lang/cargo/pull/9030/commits/d9e97c26efa99d52d30961a256811c43f7187433#diff-3bc7476e0b88b61ea4e3bd46743021ef33c927f3052c1f7800bc8aa053627b31)?)

#### Unresolved issues

- [ ] Define interaction with features: in workspaces where crate X is built for architecture A and depends on crate Z with feature F, and crate B is built for architecture B and depends on crate Z with feature G, we want the A version of Z to be built with F only and the B version of Z to be built with G only; but the current behavior is to build both the A and B versions of Z with both F and G
- [ ] Make sure things work correctly with artifact dependencies, that could introduce kinds across the build graph

#### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however not meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.
```

(I also guess a separate issue should be opened for the first unresolved issue at least, so as to be able to track it properly)

bors commented:
:umbrella: The latest upstream changes (presumably #9290) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Ekleog commented:
I've just squash-rebased onto a newer master, sorry for changing the history but always having to handle merge conflicts due to changes that I had done and undone in later commits was getting cumbersome — hopefully github will still show the commit history by clicking the links in the “force-pushed” line :)

bors commented:
:umbrella: The latest upstream changes (presumably #9298) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Ekleog commented:
@alexcrichton Could you give another look at this PR? I think I handled your last comments, so hopefully it should be ready to go :)

alexcrichton commented on crates/cargo-test-support/Cargo.toml:
```diff
@@ -5,6 +5,8 @@ authors = ["Alex Crichton <alex@alexcrichton.com>"]
 license = "MIT OR Apache-2.0"
 edition = "2018"
 
+build = "build.rs"
```
It's ok to omit this (since it's the default)

Ekleog replied:
True, thank you for the pointer :)

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -702,13 +697,56 @@ impl RustcTargetData {
+            if let CompileKind::Target(target) = kind {
+                match res.target_config.entry(target) {
+                    std::collections::hash_map::Entry::Occupied(_) => (),
+                    std::collections::hash_map::Entry::Vacant(place) => {
```
I think it's ok to use `contains_key` followed by `insert` here to avoid the verbose-ness of the `Entry`

Ekleog replied:
Right, these maps will probably stay small so the overhead should be minimal :)

alexcrichton commented on src/cargo/util/toml/mod.rs:
```diff
@@ -1313,9 +1318,24 @@ impl TomlManifest {
+            .default_target
+            .as_ref()
+            .map(|t| CompileTarget::new(&*t))
+            .transpose()? // TODO: anyhow::Context isn't imported yet so I guess .context() isn't the right way to do it?
```
I think it's safe to remove this `TODO`?

Ekleog replied:
Hmm, it's true that I had totally forgotten about this TODO, sorry! It was there because I was thinking that there may be a want to add details to the error message, but I have basically no knowledge of the cargo error handling story yet.

I've removed this TODO for now, feel free to let me know if you think there should be an equivalent to `.context()` here to give a hint to the user that the issue comes from a mis-spelled `default_target` (if I understood correctly the error conditions of `CompileTarget::new`) :)

alexcrichton reviewed (commented):
Oops sorry this fell under my radar! I've lost a bit of context since I last looked at this, but as-is this looks good to me. I suspect there's at least one or two things that will need to be handled later on, but for now this seems fine. Want to go ahead and file a tracking issue and fill that in in the unstable documentation? With some small updates below I think this should be good to merge.

Ekleog commented:
Thank you for your feedback! And no problem about this falling under your radar, it fell under mine for a long time before that :sweat_smile: 

The review comments should have been handled, I rebased and fixed a compile error that appeared since the last rebase, CI passed, and I opened https://github.com/rust-lang/cargo/issues/9406 for the tracking issue. I'm just not clear about the unstable documentation, as AFAIU the rust unstable book is only for rustc and this PR [already has changes](https://github.com/rust-lang/cargo/pull/9030/files#diff-e35cef716988e9f7122a9c90479aa9204e61d1f41b094c0d183a44e0ca271eaa) to something that may be the unstable documentation?

alexcrichton commented:
Ok thanks! This looks good to me so I'll approve

@bors: r+

For the tracking issue I'll fill in some details.

bors commented:
:pushpin: Commit fcd761729e9d7853ba8e30674f9d9a45671d1ed9 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton fcd761729e9d7853ba8e30674f9d9a45671d1ed9 -->
<!-- homu: {"type":"Approved","sha":"fcd761729e9d7853ba8e30674f9d9a45671d1ed9","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit fcd761729e9d7853ba8e30674f9d9a45671d1ed9 with merge d1c0a9d9ca31ec1b861940b30c1a660dfa2e3b56...
<!-- homu: {"type":"BuildStarted","head_sha":"fcd761729e9d7853ba8e30674f9d9a45671d1ed9","merge_sha":"d1c0a9d9ca31ec1b861940b30c1a660dfa2e3b56"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/2439791423)
Approved by: alexcrichton
Pushing d1c0a9d9ca31ec1b861940b30c1a660dfa2e3b56 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/2439791423"},"merge_sha":"d1c0a9d9ca31ec1b861940b30c1a660dfa2e3b56"} -->

Ekleog commented:
Looks like this landed without issues. Awesome, thank you! :heart:

leapofazzam123 commented:
How can I use this?

ewoolsey commented:
Can someone please explain how to use this feature? I can't find any documentation

weihanglo commented:
Normally you can find the documentation of an unstable feature from ["Unstable Features"](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html) chapter in The Cargo Book. The usage is [documented there](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#per-package-target) as well. 

For more usages, you may also want to take a look at [how cargo tests this feature](https://github.com/rust-lang/cargo/blob/0b84a35c2c7d70df4875a03eb19084b0e7a543ef/tests/testsuite/cross_compile.rs#L156-L359), or keep an eye on [the tracking issue](https://github.com/rust-lang/cargo/issues/9406) to see how people use it.
